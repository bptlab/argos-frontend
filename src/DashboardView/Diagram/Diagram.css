.DiagramWrapper {
  height: 33%;
  display: flex;
  width: 100%;
  padding-top: 40px;
  justify-content: space-around; }

.innerDiagramWrapper {
  height: 300px; }
