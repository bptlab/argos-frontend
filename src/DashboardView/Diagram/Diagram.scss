.DiagramWrapper {
  height: 33%;
  display: block;
  width: 100%;
}

.innerDiagramWrapper {
  display: block;
  margin:0 auto;
  height:100%;
  width: 20%;
}