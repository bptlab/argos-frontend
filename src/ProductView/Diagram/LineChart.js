import React, { Component } from 'react';
import Chart from "chart.js";
import {argosConfig} from '../../config/argosConfig.js';

class LineChart extends Component {
    
    static sortEventsByTime(elementA, elementB, timeStampAttribute) {
        const dateA = new Date(elementA[timeStampAttribute]);
        const dateB = new Date(elementB[timeStampAttribute]);
        return dateA - dateB;
    }
    
    buildChartData() {
        const dataset = [];
        this.props.eventData.forEach((eventDataContainer) => {
            const evenTypeDiagramData = this.buildChartDataset(eventDataContainer.eventType, eventDataContainer.events);
            dataset.push(evenTypeDiagramData);
        });
        return dataset;
    }
    buildChartDataset(eventType, events) {
        const dataContainer = [];
        const timeStampAttribute = eventType.timestampAttributeName;
        let counter = 0;
        const sortedEvents = events;
        sortedEvents.sort((elementA, elementB) => {
            return LineChart.sortEventsByTime(elementA, elementB, timeStampAttribute);
        });
        sortedEvents.forEach(function (event) {
            counter = counter + 1;
            dataContainer.push({
                x: event[timeStampAttribute].split("T")[0],
                y: counter
            });
        });
        return({
            label: eventType.name,
            fill: true,
            lineTension: 0.1,
            backgroundColor: "rgba(0, 78, 100, 0.5)",
            borderColor: "rgba(0, 78, 100, 1)",
            pointBorderWidth: 5,
            pointHoverRadius: 8,
            pointRadius: 1,
            pointHitRadius: 10,
            data: dataContainer
        });
            
    }

    componentDidUpdate() {
        const charWrapper = this.refs.canvas;
        if (charWrapper) {
            const chartContext = charWrapper.getContext('2d');
            const chartConfig = {
                type: argosConfig.kindOfDetailChart,
                data: {
                    datasets: this.buildChartData()
                },
                options: {
                    responsive: true,
                    scales: {
                        xAxes: [{
                            type: 'time',
                            time: {
                                displayFormats: {
                                    quarter: 'YYYY-MM-DD'
                                }
                            }
                        }]
                    }
                }
            };
            const chart = new Chart(chartContext, chartConfig);
            this.state = {
                chart: chart
            };
        }
    }
    
    render() {
        return (
            <div className="line-chart container">
                <canvas ref="canvas" id="lineChart" height="80"/>
            </div>
        );
    }
} export default LineChart;