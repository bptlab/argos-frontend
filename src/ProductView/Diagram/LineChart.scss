.line-chart {
	margin-bottom: $element-margin-bottom * 2;
}