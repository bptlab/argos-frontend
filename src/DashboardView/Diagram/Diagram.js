import React, { Component } from 'react';
import PieChart from './PieChart/PieChart.js'
import './Diagram.css';

class Diagram extends Component {
    constructor() {
        super();
        this.chartData = [];
        this.chartLabels = [];
    }

    fetchChartData() {
        for(let i = 0; i < this.props.products.length; i++) {
            const productState = this.props.products[i].state;
            if(this.chartLabels.indexOf(productState) > -1) {
                this.chartData[this.chartLabels.indexOf(productState)] += 1;
            } else {
                this.chartLabels.push(productState);
                this.chartData[this.chartLabels.indexOf(productState)] = 1;
            }
        }
    }

    render() {
        this.fetchChartData();
        return (
            <div className="container">
                <div className="DiagramWrapper">
                    <div className="innerDiagramWrapper">
                        <PieChart
                            chartData={this.chartData}
                            chartLabels={this.chartLabels}/>
                    </div>
                </div>
            </div>
        );
    }
}
export default Diagram;
