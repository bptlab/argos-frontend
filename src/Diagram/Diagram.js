import React, { Component } from 'react';
import './Diagram.css';
import PieChart from '../PieChart/PieChart.js'

class Diagram extends Component {

    constructor() {
        super();
        this.chartData = [];
        this.chartLabels = [];
    }


    fetchChartData() {
        for(let index in this.props.products) {
            if(this.chartLabels.indexOf(this.props.products[index].state) > -1) {
                this.chartData[this.chartLabels.indexOf(this.props.products[index].state)] += 1;
            } else {
                this.chartLabels.push(this.props.products[index].state);
                this.chartData[this.chartLabels.indexOf(this.props.products[index].state)] = 1;
            }
        }
    }

    render() {
        this.fetchChartData();
        return (
            <div className="DiagramWrapper">
                <div className="innerDiagramWrapper">
                    <PieChart chartData={this.chartData} chartLabels={this.chartLabels} />
                </div>
            </div>
        );
    }
}

export default Diagram;
