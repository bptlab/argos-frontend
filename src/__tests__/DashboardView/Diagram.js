import React from 'react';
import Diagram from '../../DashboardView/Diagram/Diagram.js';
import renderer from 'react-test-renderer';
import TestData from '../testData/products.js'
let instance;

test("Correct drawing of Diagram", () => {
    const component = renderer.create(
        <Diagram 
            ref={(child) => {instance = child}}
            products={TestData.PRODUCTS}
            onStateExcludeInput={() => {/* do nothing in case of test */}}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Correct extraction of data for displaying", () => {
    const simpleProducts = [{
        state: "WARNING"
    },{
        state: "WARNING"
    }];
    instance.chartData = [];
    instance.chartLabels = [];
    instance.fetchChartData(simpleProducts);
    expect(instance.chartLabels.length).toEqual(1);
    expect(instance.chartData.length).toEqual(1);
    expect(instance.chartLabels[0]).toEqual("WARNING");
    expect(instance.chartData[0]).toEqual(2);
});