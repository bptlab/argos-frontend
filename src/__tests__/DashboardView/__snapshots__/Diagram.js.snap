exports[`test Correct drawing of Diagram 1`] = `
<div
  className="container">
  <div
    className="diagram">
    <div
      className="row">
      <div
        className="col-xs-12 offset-sm-3 col-sm-6 offset-lg-4 col-lg-4 offset-uhd-5 col-uhd-2">
        <canvas
          id="pieChart" />
      </div>
    </div>
  </div>
</div>
`;
